portability-spi-gateway/src/main/java/org/dataportabilityproject/spi/gateway/extension/GatewayExtension.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2018 The Data-Portability Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.dataportabilityproject.spi.gateway.extension;
+
+import org.dataportabilityproject.api.launcher.AbstractExtension;
+import org.dataportabilityproject.spi.gateway.auth.AuthServiceProvider;
+import org.dataportabilityproject.spi.gateway.provider.ServiceProviderInfo;
+
+import java.util.List;
+
+/**
+ * Gateway extensions implement this interface to be loaded in a gateway process.
+ */
+public interface GatewayExtension extends AbstractExtension {
+
+    /**
+     * Returns service provider information supported by this extension.
+     */
+    List<ServiceProviderInfo> getServiceProviderInfos();
```

Comment:
So the result of this would then later be used to initialize the guice injections? 
I think this aligns well with the work I was starting for implementing the Auth and Transfer ServiceProviderRegistry classes. I have already started implementation and was trying to figure out how the registry would obtain the map of providers (similar to what we do [here](https://github.com/google/data-portability/blob/377b38e9d3997ca42356d9615b7cb4147e614644/portability-core/src/main/java/org/dataportabilityproject/ServiceProviderRegistry.java#L52))

Reply:
Yes I think the injectors could be initialized with those or they could be manually registered with the instances produced by the injectors. We may need to use a hierarchy of injectors if there are timing issues, e.g. extensions require a set of services to initialize. In this case we could use an hierarchy of Injector instances to handle this.